subsys/dfu/dfu_target/src/dfu_target_stream.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <logging/log.h>
+#include <storage/stream_flash.h>
+#include <stdio.h>
+
+LOG_MODULE_REGISTER(dfu_target_stream, CONFIG_DFU_TARGET_LOG_LEVEL);
+
+static struct stream_flash_ctx stream;
+static const char *current_id;
+
+#ifdef CONFIG_DFU_TARGET_STREAM_SAVE_PROGRESS
+
+static char current_name_key[12];
+
+#define MODULE "dfu"
+#define DFU_STREAM_OFFSET "stream/offset"
+#include <settings/settings.h>
+/**
+ * @brief Store the information stored in the stream_flash instance so that it
+ *        can be restored from flash in case of a power failure, reboot etc.
+ */
+static int store_progress(void)
+{
+	int err;
+	size_t bytes_written = stream_flash_bytes_written(&stream);
+
+	err = settings_save_one(current_name_key, &bytes_written,
+			sizeof(bytes_written));
+
+	if (err) {
+		LOG_ERR("Problem storing offset (err %d)", err);
+		return err;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Function used by settings_load() to restore the stream_flash ctx.
+ *	  See the Zephyr documentation of the settings subsystem for more
+ *	  information.
+ */
+static int settings_set(const char *key, size_t len_rd,
+			settings_read_cb read_cb, void *cb_arg)
+{
+	if (!strcmp(key, current_id)) {
+		ssize_t len = read_cb(cb_arg, &stream.bytes_written,
+				      sizeof(stream.bytes_written));
+		if (len != sizeof(stream.bytes_written)) {
+			LOG_ERR("Can't read stream.bytes_written from storage");
+			return len;
+		}
+	}
+
+	return 0;
+}
+#endif /* CONFIG_DFU_TARGET_STREAM_SAVE_PROGRESS */
+
+struct stream_flash_ctx *dfu_target_stream_get_stream(void)
+{
+	return &stream;
+}
+
+int dfu_target_stream_init(const char *id, const struct device *fdev,
+			   uint8_t *buf, size_t len, size_t offset, size_t size,
+			   stream_flash_callback_t cb)
```

Comment:
IMO more than 4 arguments makes the function unwieldy. Why not combine them into an init struct that can be passed in here? Keeps the arg list clean and adds some future-proofing to the API.

Might also be more personal preference, so not a blocker by any means.

Reply:
`dfu_target_stream_init` will never be exposed and called directly I would assume. Sitting behind the publicly `dfu_target_init()` which has fewer arguments.

But you are correct that maintaining it can be a bit hard. So we should maybe consider it.

Reply:
Added init struct. Did some weird stuff in the test to avoid having to expand that list declaration all the time, please check it.